bots/csr/src/main/java/org/openjdk/skara/bots/csr/CSRBot.java

```diff
@@ -105,28 +92,55 @@ private String describe(PullRequest pr) {
                     if (resolution == null || resolution.isNull()) {
-                        log.info("CSR issue resolution is null for " + describe(pr) + ", not removing CSR label");
+                        if (!pr.labelNames().contains(CSR_LABEL)) {
+                            log.info("CSR issue resolution is null for " + describe(pr) + ", adding the CSR label");
+                            pr.addLabel(CSR_LABEL);
+                        } else {
+                            log.info("CSR issue resolution is null for " + describe(pr) + ", not removing the CSR label");
+                        }
                         continue;
                     }
                     var name = resolution.get("name");
                     if (name == null || name.isNull()) {
-                        log.info("CSR issue resolution name is null for " + describe(pr) + ", not removing CSR label");
+                        if (!pr.labelNames().contains(CSR_LABEL)) {
+                            log.info("CSR issue resolution name is null for " + describe(pr) + ", adding the CSR label");
+                            pr.addLabel(CSR_LABEL);
+                        } else {
+                            log.info("CSR issue resolution name is null for " + describe(pr) + ", not removing the CSR label");
+                        }
                         continue;
                     }
 
                     if (csr.state() != Issue.State.CLOSED) {
-                        log.info("CSR issue state is not closed for " + describe(pr) + ", not removing CSR label");
+                        if (!pr.labelNames().contains(CSR_LABEL)) {
+                            log.info("CSR issue state is not closed for " + describe(pr) + ", adding the CSR label");
+                            pr.addLabel(CSR_LABEL);
+                        } else {
+                            log.info("CSR issue state is not closed for " + describe(pr) + ", not removing the CSR label");
+                        }
                         continue;
                     }
 
                     if (!name.asString().equals("Approved")) {
-                        log.info("CSR issue resolution is not 'Approved' for " + describe(pr) + ", not removing CSR label");
+                        if (name.asString().equals("Withdrawn")) {
+                            // This condition is necessary to prevent the bot from adding the CSR label again.
+                            // And the bot can't remove the CSR label automatically here.
+                            // Because the PR author with the role of Committer may withdraw a CSR that
+                            // a Reviewer had requested and integrate it without satisfying that requirement.
+                            log.info("CSR closed and withdrawn for " + describe(pr) + ", not removing CSR label");
+                        } else if (!pr.labelNames().contains(CSR_LABEL)) {
```

Comment:
I'm not sure about this case. For this to realistically happen, a CSR would have been filed and withdrawn before the PR was created. I don't think we should automatically add the label in this situation. @kevinrushforth what do you think?

If the label is already there, then we should certainly leave it there however.

Reply:
I agree that we need not (should not) add a label in this case. I also agree that if there is a label we must leave it.

This will handle the case where a PR author thinks a CSR might be needed, creates a CSR, withdraws it, creates the PR, and then later a Reviewer indicates that a CSR is needed. In that case, the bot should add the label and not auto-remove it.

Reply:
@erikj79 @kevinrushforth Thanks for the comments. It seems that my current code meets your comments. Did I misunderstand anything?

Please see the code below. When the CSR has been withdrawn, the bot only prints the information and doesn't add the csr label automatically. 

```
if (!name.asString().equals("Approved")) {
    if (name.asString().equals("Withdrawn")) {
        log.info("CSR closed and withdrawn for " + describe(pr) + ", not removing CSR label");  // <----here
    } else if (!pr.labelNames().contains(CSR_LABEL)) {
        log.info("CSR issue resolution is not 'Approved' for " + describe(pr) + ", adding the CSR label");
        pr.addLabel(CSR_LABEL);
    } else {
        log.info("CSR issue resolution is not 'Approved' for " + describe(pr) + ", not removing the CSR label");
    }
    continue;
}
```


Maybe I should change the output message from `"CSR closed and withdrawn for " + describe(pr) + ", not removing CSR label"` to `"CSR closed and withdrawn for " + describe(pr) + ", not revising (not adding or not removing) CSR label"`.

Reply:
You are right, I misread your code.